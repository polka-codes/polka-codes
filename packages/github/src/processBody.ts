import remarkGfm from 'remark-gfm'
import remarkParse from 'remark-parse'
import type { Root } from 'remark-parse/lib'
import remarkStringify from 'remark-stringify'
import { type Transformer, unified } from 'unified'
import { SKIP, visit } from 'unist-util-visit'

const polkaCodesRegex = /<!-- polka-codes-start -->[\s\S]*?<!-- polka-codes-end -->/g
const htmlCommentRegex = /<!--[\s\S]*?-->/g

const transformer: Transformer<Root> = (tree) => {
  visit(tree, 'html', (node, index, parent) => {
    const updated = node.value.replace(htmlCommentRegex, '')
    if (updated.length === 0 && parent && index !== undefined) {
      parent.children.splice(index, 1)
      return [SKIP, index]
    }
    node.value = updated
  })
}

const plugin = () => {
  return transformer
}

export const processBody = async (body: string) => {
  const processedBody = body.replace(polkaCodesRegex, '')
  const output = await unified().use(remarkParse).use(remarkStringify).use(remarkGfm).use(plugin).process(processedBody)
  return output.toString()
}
