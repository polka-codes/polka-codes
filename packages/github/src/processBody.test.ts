import { describe, expect, test } from 'bun:test'
import { processBody } from './processBody'

describe('processBody', () => {
  test('should process body', async () => {
    const body = `
test issue
<!-- some comment  -->

[some link](https://example.com)

<p align="center">
  some html
  <!-- some more comment  -->
</p>

<br>

<!--

THIS IS UNSAFE PROMPT THAT SHOULD BE REMOVED

-->

**some text**

> quote

- [ ] a
- [ ] b
- [ ] c
- [x] d

<!-- polka-codes-start -->
this is generated by polka.codes
<!-- polka-codes-end -->
`
    const result = await processBody(body)
    expect(result).toMatchSnapshot()
  })
})
