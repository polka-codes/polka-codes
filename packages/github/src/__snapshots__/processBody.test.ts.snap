// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`processBody should process body 1`] = `
"test issue

[some link](https://example.com)

<p align="center">
  some html
  
</p>

<br>

**some text**

> quote

* [ ] a
* [ ] b
* [ ] c
* [x] d
"
`;
